File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/matcher/AbstractMatcher.java
package com.microsoft.recognizers.text.matcher;

public abstract class AbstractMatcher<T> implements IMatcher<T> {
    abstract void insert(Iterable<T> value, String id);

    protected void batchInsert(Iterable<T>[] values, String[] ids) {
        if (values.length != ids.length) {
            throw new IllegalArgumentException();
        }

        for (int i = 0; i < values.length; i++) {
            insert(values[i], ids[i]);
        }
    }

    boolean isMatch(Iterable<T> queryText) {
        return find(queryText) != null;
    }
}

File: Python/libraries/recognizers-text/recognizers_text/matcher/abstract_matcher.py
from abc import ABC, abstractmethod
from typing import List
from recognizers_text.model import ModelResult


class Matcher(ABC):
    @property
    @abstractmethod
    def model_type_name(self) -> str:
        pass

    @abstractmethod
    def parse(self, source: str) -> List[ModelResult]:
        pass


class AbstractMatcher(Matcher):
    @property
    @abstractmethod
    def __init__(self, values: [str], ids: [str]):
        pass

    @property
    @abstractmethod
    def find(self, query_text: [str]) -> [str]:
        pass

    @property
    @abstractmethod
    def insert(self, value: [str], id: str):
        pass

    @property
    @abstractmethod
    def insert(self, value: [str], id: str):
        pass

    def is_match(self, query_text: [str]):
        pass

    def batch_insert(self, query_text: [str], ids: str):
        pass

    def convert_dict_to_list(self, node):
        pass
